src/Raven.Server/Commercial/LicenseManager.cs

```diff
@@ -121,40 +161,138 @@ public void Initialize(StorageEnvironment environment, TransactionContextPool co
         }
 
-        private void ReloadLicense()
+        private int GetClusterSize()
         {
-            var license = _serverStore.LoadLicense();
-            if (license == null)
+            if (_serverStore.IsPassive())
+                return 1;
+
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
             {
-                _licenseStatus.Attributes = null;
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
-                return;
+                return _serverStore.GetClusterTopology(context).AllNodes.Count;
             }
+        }
+
+        public async Task RecalculateLicenseLimitsIfNeeded(
+            string assignedNodeTag = null, int? assignedCores = null)
+        {
+            if (_serverStore.IsLeader() == false)
+                return;
 
             try
             {
-                _licenseStatus.Attributes = LicenseValidator.Validate(license, RSAParameters);
-                _licenseStatus.Error = false;
-                _licenseStatus.Message = null;
+                _recalculateLicenseLimitsSemsphore.Wait();
+
+                var licenseLimits = _serverStore.LoadLicenseLimits() ?? new LicenseLimits();
+                var coresByNode = licenseLimits.CoresByNode ?? 
+                    (licenseLimits.CoresByNode = new Dictionary<string, int>());
+
+                using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    if (assignedNodeTag != null && assignedCores != null)
+                    {
+                        coresByNode[assignedNodeTag] = assignedCores.Value;
+                    }
+
+                    var allNodeTags = _serverStore.GetClusterTopology(context).AllNodes.Keys;
+                    var coresInUse = coresByNode.Sum(x => x.Value);
+                    if (allNodeTags.Count == coresByNode.Count &&
+                        coresInUse <= _licenseStatus.MaxCores)
+                    {
+                        return;
+                    }
+
+                    var nodesToRemove = coresByNode.Keys.Except(allNodeTags).ToList();
+                    foreach (var nodeToRemove in nodesToRemove)
+                    {
+                        coresByNode.Remove(nodeToRemove);
+                    }
+
+                    var missingNodesAssignment = new List<string>();
+                    foreach (var nodeTag in allNodeTags)
+                    {
+                        if (coresByNode.TryGetValue(nodeTag, out var _) == false)
+                            missingNodesAssignment.Add(nodeTag);
+                    }
+
+                    if (missingNodesAssignment.Count > 0)
+                    {
+                        var availableCores = _licenseStatus.MaxCores - coresInUse;
```

Comment:
I don't like the term `MaxCores` and `MaxRamInGb`, better to call is `ClusterCores` and `ClusterRamInGb`